packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
+            require(
+                IERC20(asset).transferFrom(msg.sender, address(this), amountDeposited),
+                'Could not deposit ERC20s'
+            );
+        }
+
+        holdings[asset][destination] = holdings[asset][destination].add(amountDeposited);
+        emit Deposited(destination, asset, amountDeposited, holdings[asset][destination]);
+
+        if (asset == address(0)) {
+            // refund whatever wasn't deposited.
+            uint256 refund = amount.sub(amountDeposited);
+            (bool success, ) = msg.sender.call{value: refund}(''); //solhint-disable-line avoid-low-level-calls
+            require(success, 'Could not refund excess funds');
+        }
+    }
+
     /**
      * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation
+     * @param outcomeBytes The encoded Outcome of this state channel
      * @param indices Array with each entry denoting the index of a destination to transfer funds to. An empty array indicates "all".
      */
     function transfer(
+        uint256 assetIndex, // TODO consider a uint48?
         bytes32 fromChannelId,
-        bytes calldata allocationBytes,
+        bytes memory outcomeBytes,
+        bytes32 stateHash,
+        address challengerAddress,
         uint256[] memory indices
-    ) external override {
+    ) public override returns (bytes memory) {
         // checks
         _requireIncreasingIndices(indices);
-        _requireCorrectAllocationHash(fromChannelId, allocationBytes);
+        _requireChannelFinalized(fromChannelId);
+        _requireMatchingFingerprint(
+            stateHash,
+            challengerAddress,
+            keccak256(outcomeBytes),
+            fromChannelId
+        );
+
+        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
+        Outcome.AssetOutcome memory assetOutcome = abi.decode(
+            outcome[assetIndex].assetOutcomeBytes,
+            (Outcome.AssetOutcome)
+        );
+        require(
+            assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+            '!allocation'
+        );
+        Outcome.AllocationItem[] memory allocation = abi.decode(
+            assetOutcome.allocationOrGuaranteeBytes,
+            (Outcome.AllocationItem[])
+        );
+        address asset = outcome[assetIndex].assetHolderAddress;
+
         // effects and interactions
-        _transfer(fromChannelId, allocationBytes, indices);
+        allocation = _transfer(asset, fromChannelId, allocation, indices); // update in place to newAllocation
```

Comment:
Do you know what happens in the case where the transfer leaves the allocation empty (in other words, after the transfer, the `allocationItem` array contains no more entries)?

Reply:
I've found it much easier to operate "in place" on the allocation, or to build up a copy of it with the same length. This means that "zero allocations" (e.g. `Alice: 0`) are included in the outcome and not trimmed out. 

Some "clean up" is still possible, by tracking some "safe to delete" flags. When we know the only remaining allocations are for zero, we can just delete that `AssetOutcome` entirely. That's what happens on `master`. On this branch, we don't store `AssetOutcomes` separately so we can't do that cleanup.

I think we should circle back to this question once we get the project's goals finished, and look to optimise the behaviour. For example, if all `AssetOutcomes` are safe to delete, we can delete the `statusOf` entry for the channel (see https://github.com/statechannels/statechannels/pull/3394).